include/openPMD/Series.hpp

```diff
@@ -266,39 +312,24 @@ class Series : public Attributable
      */
     void flush();
 
-    /**
-     * @brief Entry point to the reading end of the streaming API.
-     *
-     * Creates and returns an instance of the ReadIterations class which can
-     * be used for iterating over the openPMD iterations in a C++11-style for
-     * loop.
-     * Look for the ReadIterations class for further documentation.
-     *
-     * @return ReadIterations
-     */
-    ReadIterations
-    readIterations();
+OPENPMD_private:
+    static constexpr char const * const BASEPATH = "/data/%T/";
 
-    /**
-     * @brief Entry point to the writing end of the streaming API.
-     *
-     * Creates and returns an instance of the WriteIterations class which is a
-     * restricted container of iterations which takes care of
-     * streaming semantics.
-     * The created object is stored as member of the Series object, hence this
-     * method may be called as many times as a user wishes.
-     * Look for the WriteIterations class for further documentation.
-     *
-     * @return WriteIterations
-     */
-    WriteIterations
-    writeIterations();
+    struct ParsedInput;
+    using iterations_t = decltype(internal::SeriesData::iterations);
+    using iterations_iterator = iterations_t::iterator;
 
-    Container< Iteration, uint64_t > iterations;
+    internal::SeriesData * m_series;
```

Comment:
I wonder if at this point, is a raw pointer the best pattern?

Reply:
The `…Impl` classes are pure interface classes, with the explicit goal of allowing us to be generic over resource management. The reason why we need the Impl classes is because we want a unified implementation for classes that manage their resources in different ways:
* As part of the same object, pinned in one memory location, as in `SeriesInternal`.
* Managed as a shared resource with a `shared_ptr`, as in `Series`.

So, the concrete approach at resource management depends on the class deriving from `SeriesImpl` – `SeriesImpl` must support any approach. Two ways to do this:
1. Use templates to specifiy the kind of resource management to be used by `SeriesImpl`. That's what I had implemented previously. Avoids using pointers, at the cost of introducing templates, nearly doubling compile times in my tests.
2. Relieve `SeriesImpl` off the task of resource management alltogether and just use a pointer. It's the task of the deriving class to deal with resource management and ensure via RAII mechanisms that this is done in a safe manner.